File: src/main/kotlin/interview_150/array/MergeSortedArray.java
package interview_150.array;

/**
 * Difficulty : Easy
 * Problem Statement : https://leetcode.com/problems/merge-sorted-array/description/?envType=study-plan-v2&envId=top-interview-150
 */

public class MergeSortedArray {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        int i = m - 1;
        int j = n - 1;
        int k = m + n - 1;

        while (j >= 0) {
            if (i >= 0 && nums1[i] > nums2[j]) {
                nums1[k--] = nums1[i--];
            } else {
                nums1[k--] = nums2[j--];
            }
        }
    }
}

File: src/main/kotlin/two_weeks_study_plan/data_structure/data_structure_1/day_2/MergeSortedArray.kt
package two_weeks_study_plan.data_structure.data_structure_1.day_2

/**
 * Difficulty : Easy
 * Problem Statement : https://leetcode.com/problems/merge-sorted-array/?envType=study-plan&id=data-structure-i
 */

fun main() {

    val array1 = arrayOf(1, 2, 3, 0, 0, 0).toIntArray()
    val array2 = arrayOf(2, 5, 6).toIntArray()

    MergeSortedArray().merge(array1, 3, array2, 3)

    print(array1.joinToString())

}

class MergeSortedArray {
    fun merge(nums1: IntArray, m: Int, nums2: IntArray, n: Int): Unit {
        var firstArrayLast = m - 1
        var secondArrayLast = n - 1

        var left = nums1.size - 1

        while (left >= 0) {
            if (secondArrayLast < 0) break

            if (firstArrayLast >= 0 && nums1[firstArrayLast] > nums2[secondArrayLast]) {
                nums1[left] = nums1[firstArrayLast--]
            } else {
                nums1[left] = nums2[secondArrayLast--]
            }
            left--
        }
    }
}